vispy/visuals/histogram.py

```diff
@@ -3,56 +3,114 @@
 # Copyright (c) Vispy Development Team. All Rights Reserved.
 # Distributed under the (new) BSD License. See LICENSE.txt for more info.
 # -----------------------------------------------------------------------------
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Callable
 
 import numpy as np
 
 from .mesh import MeshVisual
 
+if TYPE_CHECKING:
+    from typing import TypeAlias, Literal, SupportsIndex
+    import numpy.typing as npt
+    from vispy.color.color_array import Color
+
+    BinsLike: TypeAlias = "str | SupportsIndex | npt.ArrayLike"
+    HistogramCallable: TypeAlias = Callable[
+        [npt.ArrayLike, BinsLike], tuple[npt.NDArray, npt.NDArray]
+    ]
+    Orientation: TypeAlias = Literal["h", "v"]
+
 
 class HistogramVisual(MeshVisual):
     """Visual that calculates and displays a histogram of data
 
     Parameters
     ----------
-    data : array-like
-        Data to histogram. Currently only 1D data is supported.
-    bins : int | array-like
-        Number of bins, or bin edges.
-    color : instance of Color
-        Color of the histogram.
+    data : array-like, optional
+        Data to histogram.  May be `None` on initialization, use `set_raw_data`
+        to set data after initialization.
+    bins : int | array-like | str
+        If `bins` is an int, it defines the number of equal-width
+        bins in the given range (10, by default). If `bins` is a
+        sequence, it defines a monotonically increasing array of bin edges,
+        including the rightmost edge, allowing for non-uniform bin widths.
+        May also be a string if the calc_hist function supports it.
+    color : str | Color
+        Color of the faces in the histogram mesh.
     orientation : {'h', 'v'}
         Orientation of the histogram.
+    calc_hist : Callable
+        Function that computes the histogram. Must accept two positional arguments
+        (data, bins) and return (hist_data, bin_edges). Default is numpy.histogram.
+    **kwargs : dict
+        Keyword arguments to pass to `MeshVisual`.
```

Comment:
Thoughts on removing the type definitions in the docstring? We'd have to check the generated docs, but I'm pretty sure sphinx should handle this fine and include the types in the API docs. Not sure it is needed in the docstring otherwise, right?

Reply:
i don't have any strong thoughts on that. I'd say it's a project-specific decision, and I can see arguments for and against.  but, I'm happy to do whatever you want!  if vispy is moving in a particular direction with docstrings and you want to use this PR to update this particular module, that's fine, just let me know how you want it to look :)

Reply:
@brisvag Thoughts? @tlambert03 what have you seen other projects do? It seems redundant to have you put this much work into defining the type annotations with TypeAliases and then have to type them again in the docstrings. Could you try removing them and generating the docs locally? Or if you want, push here and we can download the website artifact from CI and view it.

```
cd doc
make html_dev-pattern PATTERN=abcd
```

^ this should generate the docs but skip all the examples.

Reply:
I agree it's redundant.  but I do still think it's slightly a matter of opinion.  I think you'll find some holdouts who feel like docstrings are for humans and type annotations are for machines, and that docstrings are "sacred" in a sense... people in that camp would probably say "go ahead and use type annotations if you want, but leave my docstrings out of it".  They might additionally argue that strictly accurate type annotations can become verbose at the expense of human readability.  then there's strict DRY camp that says it should only ever be in one place, even if it possibly requires someone to look in a different place than they might be accustomed to in interactive usage. 

(*both* might argue that they don't care how your rendered HTML docs look, or whether you're using sphinx type hints to achieve the same final result on your webpage)

i kinda like the modern type-hints only approach... but more than anything this feels like a repo-wide discussion, and not something I would try to change incrementally in each PR that happens to touch a docstring.  I added the "regular" docs to match the rest of vispy, I added the type hints after your comment suggested to add them ... I think this general discussion belongs somewhere outside of this PR :)

Reply:
Sure. I think my thing is that I typically either am looking at API docs on the website or I'm telling my IDE to pull them up for me. I rarely open the docstring version of the docs in my IDE and only pull up the "show me the parameters and their types" view inline as I'm typing. I think PyCharm is smart enough to put type annotations where it needs to in these cases. Anyway...I don't think vispy is maintained well enough to do this repo-wide so I personally would be OK with it being done here as a test...assuming it renders properly in sphinx.

Reply:
No strong opinions, frankly ^^' I would only say, if you do add them to the docstrings keep them human-radable, in which case that's probably my preferred way.

Reply:
@tlambert03 Maybe that leaves it up to you then since we're asking you to do the work. I think I'd prefer type annotations only, @brisvag seems to have a slight preference for docstrings (human-readable). I don't think NOT having type annotations is a reasonable option so the choice is both or annotations only. Again, I vote for annotations only and we see how it goes with user complaints (if any).

Reply:
ok, so just so I'm clear.  you want me to add type annotations to parts of the code didn't previously have them? (like that one method in plotwidget)?  You're ok with just that single method having type annotations?  or you want me to add it to the whole module?  (this is why i find this all a bit confusing... i imagined this would be a quick PR that just moves a little existing code around to allow it to be called as a method 😄 )

Reply:
Then it looks like what we have here already will do just fine (better than anything we had before anyways :P)

Reply:
ok, will do.  I believe I added type annotations in the `__init__` in response to a previous comment ... it wasn't in the original PR, which largely aimed to match the existing code patterns.  
I'll go ahead and remove docstring types in HistogramVisual